config/v1/types_oauth.go

```diff
@@ -1,37 +1,478 @@
+	// The value represents the maximum amount of time that can occur between
+	// consecutive uses of the token. Tokens become invalid if they are not
+	// used within this temporal window. The user will need to acquire a new
+	// token to regain access once a token times out.
+	// Valid values are integer values:
+	//   x < 0  Tokens never timeout (e.g. `-1`)
+	//   x = 0  Tokens are disabled (default)
+	//   x > 0  Tokens time out if there is no activity for x seconds
+	// The current minimum allowed value for X is 300 (5 minutes)
+	// +optional
+	AccessTokenInactivityTimeoutSeconds int32 `json:"accessTokenInactivityTimeoutSeconds,omitempty"`
+}
+
+type MappingMethodType string
+
+const (
+	// MappingMethodLookup does not provision a new identity or user, it only allows identities
+	// already associated with users
+	MappingMethodLookup MappingMethodType = "lookup"
+
+	// MappingMethodClaim associates a new identity with a user with the identity's preferred
+	// username if no other identities are already associated with the user
+	MappingMethodClaim MappingMethodType = "claim"
+
+	// MappingMethodAdd associates a new identity with a user with the identity's preferred username,
+	// creating the user if needed, and adding to any existing identities associated with the user
+	MappingMethodAdd MappingMethodType = "add"
+
+	// MappingMethodGenerate finds an available username for a new identity, based on its preferred username
+	// If a user with the preferred username already exists, a unique username is generated
+	MappingMethodGenerate MappingMethodType = "generate"
+)
+
+// OAuthIdentityProvider provides identities for users authenticating using credentials
+type OAuthIdentityProvider struct {
+	// Name is used to qualify the identities returned by this provider
+	Name string `json:"name"`
+
+	// UseAsChallenger indicates whether to issue WWW-Authenticate challenges for this provider
+	UseAsChallenger bool `json:"challenge"`
+	// UseAsLogin indicates whether to use this identity provider for unauthenticated browsers to login against
+	UseAsLogin bool `json:"login"`
+
+	// MappingMethod determines how identities from this provider are mapped to users
+	// Defaults to "claim"
+	// +optional
+	MappingMethod MappingMethodType `json:"mappingMethod"`
+
+	// GrantMethod: allow, deny, prompt
+	// This method will be used only if the specific OAuth client doesn't provide a strategy
+	// of their own. Valid grant handling methods are:
+	//  - auto:   always approves grant requests, useful for trusted clients
+	//  - prompt: prompts the end user for approval of grant requests, useful for third-party clients
+	//  - deny:   always denies grant requests, useful for black-listed clients
+	// Defaults to "prompt" if not set.
+	// +optional
+	GrantMethod GrantHandlerType `json:"grantMethod"`
+
+	// IdentityProvidersConfig
+	ProviderConfig IdentityProvidersConfig `json:",inline"`
+}
+
+type IdentityProvidersConfig struct {
```

Comment:
Create a discriminator field here.

```go
Type string
```

This should match the name of the json stanza they provide only `CamelCase`, not `camelCase`.   We provide this field so that deserialization of a later version of v1 that contains a new stanza into an older version of v1 that does not contain the new stanza is easily detectable, provides a reasonable message, and keeps the intent the same (not just a validation issue of "everything is nil!"

@openshift/api-reviewers something to remember about discriminators.  They are good.

Reply:
@deads2k does 05f5a42 look right?